criu/kerndat.c

```diff
@@ -1153,29 +1155,59 @@ int kerndat_has_nspid(void)
 static int kerndat_has_nftables_concat(void)
 {
 #if defined(CONFIG_HAS_NFTABLES_LIB_API_0) || defined(CONFIG_HAS_NFTABLES_LIB_API_1)
+	pid_t pid;
 	struct nft_ctx *nft;
-	int ret = 0;
+	int ret, status;
 
-	nft = nft_ctx_new(NFT_CTX_DEFAULT);
-	if (!nft)
+	pid = fork();
+	if (pid < 0) {
+		pr_perror("Can't fork");
 		return -1;
+	}
 
-	if (NFT_RUN_CMD(nft, "create table inet CRIU-kerndat-test")) {
-		ret = -1;
-		goto nft_ctx_free_out;
+	if (pid == 0) {
+		/*
+		 * Create a separate network namespace to avoid
+		 * collisions between two CRIU instances.
+		 */
+		if (unshare(CLONE_NEWNET)) {
+			pr_perror("Unable create a network namespace");
+			exit(-1);
+		}
+
+		nft = nft_ctx_new(NFT_CTX_DEFAULT);
+		if (!nft)
+			exit(-1);
+
+		if (NFT_RUN_CMD(nft, "create table inet CRIU")) {
+			ret = -1;
+			pr_err("Can't create nftables table\n");
+			goto nft_ctx_free_out;
+		}
+
+		if (NFT_RUN_CMD(nft, "add set inet CRIU conn { type ipv4_addr . inet_service ;}"))
+			ret = false; /* kdat.has_nftables_concat = false */
+		else
+			ret = true; /* kdat.has_nftables_concat = true */
+
+		/* Clean up */
+		NFT_RUN_CMD(nft, "delete table inet CRIU");
+
+	nft_ctx_free_out:
+		nft_ctx_free(nft);
+		exit(ret);
 	}
 
-	if (NFT_RUN_CMD(nft, "add set inet CRIU-kerndat-test conn { type ipv4_addr . inet_service ;}"))
-		kdat.has_nftables_concat = false;
-	else
-		kdat.has_nftables_concat = true;
+	if (waitpid(pid, &status, 0) != pid) {
+		pr_perror("Unable to wait %d", pid);
+		return -1;
+	}
 
-	/* Clean up */
-	NFT_RUN_CMD(nft, "delete table inet CRIU-kerndat-test");
+	if (status < 0)
```

Comment:
status would never be < 0...

Let's apply some debug patch:
```
[snorch@fedora criu]$ git diff
diff --git a/criu/kerndat.c b/criu/kerndat.c
index baae1a886..cae4d5df4 100644
--- a/criu/kerndat.c
+++ b/criu/kerndat.c
@@ -1166,6 +1166,7 @@ static int kerndat_has_nftables_concat(void)
        }
 
        if (pid == 0) {
+               exit(-1);
                /*
                 * Create a separate network namespace to avoid
                 * collisions between two CRIU instances.
@@ -1203,6 +1204,7 @@ static int kerndat_has_nftables_concat(void)
                return -1;
        }
 
+       pr_err("DEBUG %d\n", status);
        if (status < 0)
                return -1;
```
And run it:
```
[snorch@fedora criu]$ sudo criu/criu check --all
Warn  (criu/kerndat.c:846): Stale /run/criu.kdat file
Error (criu/kerndat.c:1207): DEBUG 65280
Looks good but some kernel features are missing
which, depending on your process tree, may cause
dump or restore failure.
```

Exit(-1) is equivalent of exit(255) if I recall correctly.


Note: Probably it's a good idea to reuse `call_in_child_process()`.

Reply:
E.g. like this https://github.com/checkpoint-restore/criu/pull/1591/commits/aa4b8b2b4d5269a099c1a66fd280e312285ed27d Feel free to merge to your patch if you are ok with it.